drivers/frequency/admv1013/admv1013.c

```diff
@@ -0,0 +1,395 @@
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * @brief Writes data to ADMV1013 over SPI.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data value to write.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv1013_spi_write(struct admv1013_dev *dev, uint8_t reg_addr,
+		       uint16_t data)
+{
+	uint8_t buff[ADMV1013_BUFF_SIZE_BYTES];
+
+	/*
+	    |                Byte 0           |     Byte 1     |       Byte 2        |
+	    | 0 | Addr bits 5-0 | Data bit 15 | Data bits 14-7 | Data bits 6 - 0 | 0 |
+	*/
+	buff[0] = ADMV1013_SPI_WRITE_CMD | (reg_addr << 1) | (data >> 15);
+	buff[1] = data >> 7;
+	buff[2] = data << 1;
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff,
+					ADMV1013_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Reads data from ADMV1013 over SPI.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param data - Data read from the device.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv1013_spi_read(struct admv1013_dev *dev, uint8_t reg_addr,
+		      uint16_t *data)
+{
+	uint8_t buff[ADMV1013_BUFF_SIZE_BYTES];
+	int ret;
+
+	buff[0] = ADMV1013_SPI_READ_CMD | (reg_addr << 1);
+	buff[1] = 0;
+	buff[2] = 0;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff, ADMV1013_BUFF_SIZE_BYTES);
+	if(ret)
+		return ret;
+
+	/*
+	    |                Byte 0           |     Byte 1     |       Byte 2        |
+	    | 1 | Addr bits 5-0 | Data bit 15 | Data bits 14-7 | Data bits 6 - 0 | 0 |
+	*/
+	*data = ((buff[0] & NO_OS_BIT(0)) << 15 | buff[1] << 7 | buff[2] >> 1);
+
+	return 0;
+}
+
+/**
+ * @brief Update ADMV1013 register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Mask for specific register bits to be updated.
+ * @param data - Data written to the device (requires prior bit shifting).
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+int admv1013_spi_update_bits(struct admv1013_dev *dev, uint8_t reg_addr,
+			     uint16_t mask, uint16_t data)
+{
+	uint16_t read_val;
+	int ret;
+
+	ret = admv1013_spi_read(dev, reg_addr, &read_val);
+	if (ret)
+		return ret;
+
+	read_val &= ~mask;
+	read_val |= data;
+
+	return admv1013_spi_write(dev, reg_addr, read_val);
+}
+
+/**
+ * @brief Update Mixer Gate Voltage.
+ * @param dev - The device structure.
+ * @return Returns 0 in case of success or negative error code otherwise.
+ */
+static int admv1013_update_mixer_vgate(struct admv1013_dev *dev)
+{
+	unsigned int mixer_vgate;
+
+	if (dev->vcm_uv < 1800000)
+		mixer_vgate = (2389 * dev->vcm_uv / 1000000 + 8100) / 100;
+	else if (dev->vcm_uv > 1800000 && dev->vcm_uv < 2600000)
+		mixer_vgate = (2375 * dev->vcm_uv / 1000000 + 125) / 100;
```

Comment:
would it make sense here to use macros instead of magic values?

Reply:
Well, I can invent some macros, but since the datasheet contains only the "magic" values without any further explanation, I decided to stick to that approach.


![image](https://user-images.githubusercontent.com/26061529/165726256-da047838-1cb5-4e57-9098-dda163603a02.png)
![image](https://user-images.githubusercontent.com/26061529/165726569-e0c1e239-6d05-4c55-b258-468cced65e14.png)

Any suggestions are welcome :)

Reply:
Same for mixer_vgate:
![image](https://user-images.githubusercontent.com/26061529/165727162-4e5efaa0-99c0-4b51-a83d-9f7293948a60.png)

Reply:
maybe something like LO_BAND_#_#_TO_#_#_GHZ? e.g. with value: LO_BAND_8_62_TO_10_25_GHZ

Reply:
got it, will add an enum for the quad_filter modes then :)

Reply:
and for mixer gate, maybe create directly a macro with the formula?
something like:
MIXER_GATE_0_to_1_8_V(x) ((2389 * x/ 1000000 + 8100) / 100)
MIXER_GATE_1_8_to_2_6_V(x) ((2375 * x/ 1000000 + 125) / 100)

Reply:
Sounds good. Thanks.